### PR TITLE
[CHORE ts] A couple of dabblings to start us down the road

### PR DESCRIPTION
**Pull request by runspired**
// ... please delete this default text after reading

If this is your first PR to `ember-data`, you may want to read our [Contributor Guide](https://github.com/emberjs/data/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: addon/-private/system/internal-model-map.ts [hunk 1]
```diff
@@ -24,19 +37,19 @@ export default class InternalModelMap {
...
     return this._idToModel[id];
   }
 
-  has(id) {
+  has(id: string): boolean {
     return !!this._idToModel[id];
   }
 
-  get length() {
+  get length(): number {
```
Comment (mike-north) [inline addon/-private/system/internal-model-map.ts]: If you want to use tslint to enforce explicit return types, the rule looks like

```json
"rules": {
  "typedef": [true, "call-signature"]
}
```

File: types/dummy/index.d.ts [hunk 2]
```diff
@@ -3,3 +3,25 @@
  * ember-data. They will not be shipped to consumers. Ember-data still relies
  * on some private Ember APIs -- those should be defined here as we encounter them.
  */
+declare module '@ember/debug' {
+  export function assert(msg: string, test: any): boolean;
```
Comment (igorT) [inline types/dummy/index.d.ts]: should test be `Function | boolean`

Comment (mike-north) [reply types/dummy/index.d.ts]: @igorT I usually rely on the API docs as the source of truth, and they don't indicate that a predicate can be passed in.
https://www.emberjs.com/api/ember/3.4/functions/@ember%2Fdebug/assert

I could see where passing a predicate could be confusing
```ts
assert('Problem with toString', obj.toString);
```
Does this ☝️ mean "the value of `.toString` was falsy", or "the value returned by `obj.toString()` was falsy"?


RE: whether the types should be imported vs. explicitly stated w/in ember-data, `assert` should be provided by the `@types/ember__debug` package. No need for it to be defined in this project.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/549360d30e796ac5635e019cb276e8af9933ff71/types/ember__debug/index.d.ts#L10

Comment (runspired) [reply types/dummy/index.d.ts]: I'll look into whether I can import this type correctly, it wasn't resolving with the installed ember types but perhaps I just need to add the debug module.

File: addon/-private/system/internal-model-map.ts (hunk 1 above)
Comment (mike-north) [reply addon/-private/system/internal-model-map.ts]: anything in the "rules" object applies only to TS, and there's a "jsRules" for JS

Comment (runspired) [reply addon/-private/system/internal-model-map.ts]: ooooooh :D then yes, I'll turn that on :D

File: types/dummy/index.d.ts (hunk 2 above)
Comment (mike-north) [reply types/dummy/index.d.ts]: @types/ember has @types/ember__debug as a dependency. Should work.

File: addon/-private/system/internal-model-map.ts [hunk 3]
```diff
@@ -12,7 +19,13 @@ import InternalModel from './model/internal-model';
  @private
  */
 export default class InternalModelMap {
-  constructor(modelName) {
+  private _idToModel: IDDict;
+  private _models: InternalModel[];
+  private _metadata: MetaDict | null;
```
Comment (mike-north) [inline addon/-private/system/internal-model-map.ts]: It makes sense to move the initializations in the constructor below up here.

```ts
  private _idToModel: IDDict = Object.create(null);
  private _models: InternalModel[] = [];
  private _metadata: MetaDict | null = null;

  constructor(public modelName: string) {}
```

**Pull request merged**